accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+                                 AccessControlType.Allow)
+    );
+
+    Directory.CreateDirectory(loggingDirectory, rules);
+
+    return loggingDirectory;
+}
+```
+
+### Add platform-specific code without becoming platform-specific
+
+After reviewing the code she realizes that this code only makes sense for cases
+where the application is installed on Windows. When running on iOS and Android,
+the machine isn't in shared environment so restricting the log access doesn't
+make much sense. She modifies the code accordingly:
+
+```C#
+private static string GetLoggingPath()
+{
+    var appDataDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+    var loggingDirectory = Path.Combine(appDataDirectory, "Fabrikam", "AssetManagement", "Logging");
+
+    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+    {
+        // Just create the directory
+        Directory.CreateDirectory(loggingDirectory);
+    }
+    else
+    {
+        // Create the directory and restrict access using Windows
+        // Access Control Lists (ACLs).
+
+        var rules = new DirectorySecurity();
+        rules.AddAccessRule(
+            new FileSystemAccessRule(@"fabrikam\log-readers",
+                                     FileSystemRights.Read,
+                                     AccessControlType.Allow)
+        );
+        rules.AddAccessRule(
+            new FileSystemAccessRule(@"fabrikam\log-writers",
+                                     FileSystemRights.FullControl,
+                                     AccessControlType.Allow)
+        );
+
+        Directory.CreateDirectory(loggingDirectory, rules);
+    }
+
+    return loggingDirectory;
+}
+```
+
+After that, the diagnostic disappears automatically.
+
+## Requirements
+
+### Goals
+
+* Developers get diagnostics when they inadvertently use platform-specific APIs.
+    - All diagnostics are provided on the command line as well as in the IDE
+    - The developer doesn't have to resort to explicit suppressions to indicate
+      intent. Idiomatic gestures, such as guarding the call with a platform
+      check or marking a method as platform-specific, automatically suppress
+      these diagnostics. However, explicit suppressions via the usual means
+      (`NoWarn`, `#pragma`) will also work.
+* The analyzer helps library authors to express their intent so that their
+  consumers can also benefit from these diagnostics.
+    - Library authors can annotate their platform-specific APIs without taking a
+      dependency.
+    - Library consumers don't have to install a NuGet package to get diagnostics
+      when using platform-specific APIs
+
+### Non-Goals
+
+* Modelling partially portable APIs (that is APIs are applicable to more than
+  one platform but no all platforms). Those should Be modeled as capability
+  APIs, as as `RuntimeInformation.IsDynamicCodeSupported` for ref emit. This
+  would be a separate analyzer.
```

Comment:
How are we planning to handle APIs that work everywhere but only for a subset of uses?  For example, `MemoryMapedFile.CreateFromFile(string path, System.IO.FileMode mode, string mapName)` and other such overloads work fine on all OSes, as long as `mapName` is null... it'll currently only work on Windows if `mapName` is non-null.

Reply:
In this specific case, wouldn't it make more sense to add an overload without the mapName parameter and make that not work on Windows and the one with mapName only work on Windows?

Reply:
> In this specific case, wouldn't it make more sense to add an overload without the mapName parameter and make that not work on Windows and the one with mapName only work on Windows?

There are overloads without mapName.

Reply:
> How are we planning to handle APIs that work everywhere but only for a subset of uses?

1. Try to formalize the pattern and use attributes so that we can handle these cases generically (for example, by adding an attribute like `[Supported("Windows")]` that can be applied to parameters).
2. Write a custom analyzer
3. Introduce new APIs that only allow inputs hat make sure the API works everywhere, mark the old as platform-specific.
4. Ignore

My personal opinion is that we should take a look at the BCL and see how far we can stretch (1). We can then cover the tail with approaches (2) - (4).

Reply:
Taking this specific example, how does [Supported("Windows")] apply? That attribute means not null?

For a custom analyzer, that's fine, but presumably that's part of this effort, needs to be factored into consistency around diagnostic IDs, planned for, etc. There will be a non-trivial number of such cases.